controller/engine_controller.go

```diff
@@ -612,33 +620,54 @@ func (ec *EngineController) startMonitoring(e *longhorn.Engine) {
 	ec.engineMonitorMap[e.Name] = stopCh
 
 	go monitor.Run()
+
+	go func() {
+		<-monitoringRemoveCh
+		ec.engineMonitorMutex.Lock()
+		delete(ec.engineMonitorMap, e.Name)
+		ec.logger.WithField("engine", e.Name).Debug("removed the engine from ec.engineMonitorMap")
+		ec.engineMonitorMutex.Unlock()
+	}()
```

Comment:
My current solution for engine monitoring is: 
There are 3 goroutines: the main sync goroutine, the monitoring goroutine, and the stop monitoring goroutine. 
There are 2 ways to stop monitoring an engine:
1.  The main sync goroutine closes the stopCh -> the monitoring goroutine closes the monitoringRemoveCh -> the stop monitoring goroutine removes the engine RS from the `ec.engineMonitorMap` -> Done
1. The monitoring goroutine volunteer to stop monitoring engine, it closes the monitoringRemoveCh -> the stop monitoring goroutine removes the engine RS from the `ec.engineMonitorMap` -> Done

Your suggested solution for backing image monitoring is:
There are 3 goroutines: the main sync goroutine, the monitoring goroutine, and  the [notification update](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1093) goroutine
There are 2 ways to stop monitoring a backing image:
1.  The main sync goroutine closes the stopCh and removes the backing image RS from the `ec.engineMonitorMap` -> the monitoring goroutine returns ->  the notification update goroutine returns -> Done
1. The notification update goroutine volunteers to stop monitoring backing image CR, it closes the stopCh and  removes the engine RS from the `ec.engineMonitorMap`-> the monitoring goroutine returns -> Done

The suggested solution wouldn't work if the monitoring goroutine wants to volunteer to stop monitoring the backing image. For example, if something is wrong inside the monitoring goroutine while it does [pollAndUpdateBackingImageFileMap](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1147) (i.e., the backing image's owner has moved to a new node and doesn't need to be monitored on this node), the monitoring goroutine cannot volunteer to stop monitoring the backing image. If we try to call `m.monitorStopper(m.Name)` inside `pollAndUpdateBackingImageFileMap()`, it will be blocked forever because we send [stopCH <- struct{}{}](https://github.com/shuo-wu/longhorn-manager/blob/35f06eba89ae87e0ad8e67b245b74c1ee2e1ac4c/controller/backing_image_manager_controller.go#L1066) but no goroutine would take it out.

For engine monitoring, we need the stop monitoring goroutine to be able to volunteer to stop monitoring the engine if something is wrong while doing [sync](https://github.com/longhorn/longhorn-manager/blob/f12923c00c2aab6a21b52bf311393337814902c9/controller/engine_controller.go#L686). Therefore, I think the current implementation is ok.

How do you think?

Reply:
Make sense. As a receiver, the monitor goroutine cannot send a stop signal.

Reply:
And I may need a similar logic in the backing image manager controller to actively stop monitoring in case of ownerID suddenly change.

Reply:
BTW, can you rename it to avoid misunderstanding? e.g., `monitorVoluntaryStopCh`

Reply:
Thank you! I changed `monitoringRemoveCh ` to `monitorVoluntaryStopCh `